### PR TITLE
Add solvers

### LINKED ISSUE #18: Write some ODE solvers
Some forward explicit method:
- [ ] Forward euler
- [ ] Runge-Kutta (RK) 4
- [ ] RK3
Some implicit method:
- [ ] Backward euler
- [ ] Trapezoidal method

### PR DESCRIPTION
**Pull request by zqigolden**
Resolve #18

### CODE REVIEW

File: deer/fsolve_ivp.py [hunk 1]
```diff
@@ -169,3 +169,232 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
...
+            num_steps = max(int((tf - ti) / dt), 1)
+            dt = (tf - ti) / num_steps  # Recalculate dt to evenly divide the interval
+            dx = (xf - xi) / (tf - ti) * dt
+
+            for _ in range(num_steps):
+                yi, xi = self.ode_step(func, yi, xi, dt, dx, params)
+            y.append(yi)
+
+        # Stack the list of solutions into a single jax array
+        return jnp.stack(y)
```
Comment (mfkasim1) [inline deer/fsolve_ivp.py:212]: Can you please change for loops into `jax.lax.scan`? For loop will take a long time to compile with `jax.jit`.

File: deer/fsolve_ivp.py [hunk 2]
```diff
@@ -169,3 +169,232 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
...
+class GeneralODE(SolveIVPMethod):
+    """
+    Compute the solution of initial value problem with the ODE methods.
+
+    Arguments
+    ---------
+    step_size: float
+        The step size for ODE solver. If None, it will use (tpts[i] - tpts[i - 1]).
+    """
+    def __init__(self, step_size: Optional[float] = None):
```
Comment (mfkasim1) [inline deer/fsolve_ivp.py:183]: Why do we need `step_size` here? It should be given by `tpts[i] - tpts[i - 1]`. If the users need more accuracy, then they should put denser `tpts`.

File: deer/fsolve_ivp.py [hunk 3]
```diff
@@ -169,3 +169,232 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
...
+        dx: jnp.ndarray,
+        params: Any
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        y_new = yi
+        for _ in range(self.max_iter):
+            y_new_next = yi + dt * func(y_new, xi, params)
+            if jnp.linalg.norm(y_new_next - y_new) < self.tol:
+                y_new = y_new_next
+                break
+            y_new = y_new_next
```
Comment (mfkasim1) [inline deer/fsolve_ivp.py:369]: Can you do `jax.lax.scan` here as well?

File: deer/fsolve_ivp.py [hunk 4]
```diff
@@ -169,3 +169,232 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
...
+        # Initial guess for fixed-point iteration (use forward Euler as an initial guess)
+        y_new = yi + dt * func(yi, xi, params)
+        
+        # Fixed-point iteration to solve the trapezoidal equation
+        for _ in range(self.max_iter):
+            y_new_next = yi + (dt / 2.0) * (func(yi, xi, params) + func(y_new, xi + dx, params))
+            if jnp.linalg.norm(y_new_next - y_new) < self.tol:
+                y_new = y_new_next
+                break
+            y_new = y_new_next
```
Comment (mfkasim1) [inline deer/fsolve_ivp.py:397]: Can you do `jax.lax.scan` here as well?

**Review by mfkasim1 — CHANGES_REQUESTED** [review]

File: deer/fsolve_ivp.py (hunk 4 above)
Comment (zqigolden) [reply deer/fsolve_ivp.py]: Loops are replaced with `jax.lax.scan` now.

File: deer/fsolve_ivp.py (hunk 3 above)
Comment (zqigolden) [reply deer/fsolve_ivp.py]: Loops are replaced with `jax.lax.scan` now.

File: deer/fsolve_ivp.py (hunk 2 above)
Comment (zqigolden) [reply deer/fsolve_ivp.py]: The step_size parameter was originally intended to improve the performance of solving ODEs with some given data. However, it introduces an assumption that X changes linearly between time points, which can be inaccurate. Therefore, it has been removed for simplicity.

File: deer/fsolve_ivp.py (hunk 1 above)
Comment (zqigolden) [reply deer/fsolve_ivp.py]: Yes, it has replaced.

File: deer/fsolve_ivp.py [hunk 5]
```diff
@@ -165,3 +166,230 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
...
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        
+        def cond(val):
+            y_new, y_new_next, i = val
+            return jnp.logical_and(i < self.max_iter, jnp.linalg.norm(y_new_next - y_new) >= self.tol)
+        
+        def body(val):
+            y_new, _, i = val
+            y_new_next = yi + dt * func(y_new, xi, params)
+            return y_new_next, y_new_next, i + 1
```
Comment (mfkasim1) [inline deer/fsolve_ivp.py:350]: This might not converge on some functions. You can use root finder here to improve the convergence. You can see this one to get an example on how to use the rootfinder in backward euler https://github.com/machine-discovery/deer/blob/ae4a864071937a69729ab5cc447cc6bbdf53af7f/deer/fsolve_idae.py#L108-L123

Comment (zqigolden) [reply deer/fsolve_ivp.py]: Added for backward ode methods

File: deer/tests/test_deer.py [hunk 6]
```diff
@@ -413,5 +413,32 @@ def dae_pendulum(vrdot: jnp.ndarray, vr: jnp.ndarray, t: jnp.ndarray, params) ->
...
+
+    # Compute reference solution using DEER method
+    yt_deer = solve_ivp(sample_func, y0, xinp, params, tpts, method=solve_ivp.DEER())
+
+    # Compute solution using the parameterized method
+    yt_method = solve_ivp(sample_func, y0, xinp, params, tpts, method=method)
+
+    # Compare results
+    assert jnp.allclose(yt_deer.value, yt_method.value, atol=atol)
+
```
[the remark this answers is not in the document]
Comment (zqigolden) [reply deer/tests/test_deer.py:442]: Grad check added

File: deer/fsolve_ivp.py [hunk 7]
```diff
@@ -165,3 +166,230 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
...
+
+
+class GeneralBackwardODE(GeneralODE):
+    """
+    Compute the solution of initial value problem with Backward ODE methods.
+
+    Arguments
+    ---------
+    step_size: float
+        The step size for ODE solver. If None, it will use (tpts[i] - tpts[i - 1]).
```
[the remark this answers is not in the document]
Comment (zqigolden) [reply deer/fsolve_ivp.py]: Removed

File: deer/fsolve_ivp.py [hunk 8]
```diff
@@ -173,3 +174,225 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
...
+        # Define the main function to cover all time intervals
+        def scan_step(carry, i):
+            _, success = carry
+            def success_fn(carry, i):
+                yi, _ = carry
+                xi = xinp[i-1]
+                dx = xinp[i] - xinp[i-1]
+                dt = tpts[i] - tpts[i-1]
+                yn, success = self.ode_step(func, yi, xi, dt, dx, params)
+                return yn, success
```
Comment (mfkasim1) [inline deer/fsolve_ivp.py:196]: Seems like `xinp` and `tpts` variables are coming from outside this function and that would make this function unpure. JAX has a strong obsession with pure function (see [this](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions) for example), so can you make it a pure function?

**Review by mfkasim1 — CHANGES_REQUESTED** [review]

File: deer/fsolve_ivp.py (hunk 8 above)
Comment (zqigolden) [reply deer/fsolve_ivp.py]: Fixed with [make function purer](https://github.com/machine-discovery/deer/pull/26/commits/da37873b367a24900a6a24876e7559b26c44e76c)
